app/src/main/java/org/pragyan/dalal18/fragment/ReferAndEarnFragment.kt

```diff
@@ -0,0 +1,141 @@
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.TextView
+import android.widget.Toast
+import androidx.core.content.ContextCompat
+import androidx.fragment.app.Fragment
+import androidx.lifecycle.lifecycleScope
+import com.google.android.material.snackbar.Snackbar
+import dalalstreet.api.DalalActionServiceGrpc
+import dalalstreet.api.actions.GetReferralCodeRequest
+import dalalstreet.api.actions.GetReferralCodeResponse
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import org.pragyan.dalal18.R
+import org.pragyan.dalal18.dagger.ContextModule
+import org.pragyan.dalal18.dagger.DaggerDalalStreetApplicationComponent
+import org.pragyan.dalal18.databinding.FragmentReferAndEarnBinding
+import org.pragyan.dalal18.utils.ConnectionUtils
+import org.pragyan.dalal18.utils.Constants
+import org.pragyan.dalal18.utils.viewLifecycle
+import javax.inject.Inject
+
+class ReferAndEarnFragment : Fragment() {
+
+    private var binding by viewLifecycle<FragmentReferAndEarnBinding>()
+
+    @Inject
+    lateinit var actionServiceBlockingStub: DalalActionServiceGrpc.DalalActionServiceBlockingStub
+
+    @Inject
+    lateinit var preferences: SharedPreferences
+
+    private var referralCodeDialog: AlertDialog? = null
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+        binding = FragmentReferAndEarnBinding.inflate(inflater, container, false)
+
+        DaggerDalalStreetApplicationComponent.builder().contextModule(ContextModule(context!!)).build().inject(this)
+        return binding.root
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        val dialogView = LayoutInflater.from(context!!).inflate(R.layout.progress_dialog, null)
+        (dialogView.findViewById<View>(R.id.progressDialog_textView) as TextView).setText(R.string.get_referral_code)
+        referralCodeDialog = AlertDialog.Builder(context!!).setView(dialogView).setCancelable(false).create()
+
+        binding.shareButton.setOnClickListener { shareReferralCode() }
+
+        binding.referralCodeButton.setOnClickListener { copyReferralCodeToClipboard(binding.userReferralCodeEditText.hint.toString()) }
+
+        getUserReferralCode()
+
+    }
+
+    private fun copyReferralCodeToClipboard(code: String) {
+
+        if (code.isEmpty()) {
+            Toast.makeText(context!!, "referesh your code", Toast.LENGTH_LONG).show()
+            return
+        }
+
+        val clipboardManager = activity?.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
+        val clip = ClipData.newPlainText("Copied referral code", code) as ClipData
+        clipboardManager.setPrimaryClip(clip)
+
+        Toast.makeText(context!!, "referral code copied to clipboard", Toast.LENGTH_LONG).show()
```

Comment:
```suggestion
        Toast.makeText(context!!, "Copied to clipboard", Toast.LENGTH_LONG).show()
```
And also prolly get it from `string.xml`